### PR TITLE
Implements a buffer limit for the Text widget

### PR DESCRIPTION
**Pull request by timbutler**
See issue #293 where memory and performance can degrade with a high number of lines written to the Text widget. 

This is a very simplistic implementation to limit the possible length the text buffer can grow to with the `maxContent` option. 

Default value of -1 means there's no limit and therefore behaviour should remain standard.

It has been working in our test app and allows the use of the Text widget to monitor logs (ie tail) and therefore doesn't bloat over time, but happy to adjust as required.

### CODE REVIEW

File: widgets/text/options.go [hunk 1]
```diff
@@ -174,3 +176,16 @@ func ScrollKeys(up, down, pageUp, pageDown keyboard.Key) Option {
...
 	})
 }
+
+// The default for the maximum buffer length within a content area
+// -1 sets as no limit
+const (
+	DefaultMaxContent = -1
+)
+
+// MaxContent - Limits the maximum content within text widget buffer
```
Comment (mum4k) [inline widgets/text/options.go:186]: This is the only public place where this feature is documented. Would you mind expanding the comment and explaining the behavior `Text` has when this is set?

It would also be good to specify whether we are limiting the number of runes, or the number of cells. I assume this is limiting the number of runes to be displayed and therefore the number of cells may be variable (since some unicode runes take multiple cells).

File: widgets/text/text.go [hunk 2]
```diff
@@ -109,6 +109,17 @@ func (t *Text) Write(text string, wOpts ...WriteOption) error {
 		t.reset()
 	}
 	for _, r := range text {
+		// Implement basic buffer limit to avoid excess memory issues
+		if len(t.content) > 1 && len(t.content) >= t.opts.maxContent && t.opts.maxContent > 0 {
```
Comment (mum4k) [inline widgets/text/text.go:113]: Can you help me understand the first condition `len(t.content) > 1`?

Why do we need it, or in other words wouldn't the rest of the condition be enough? I.e.

if len(t.content) >= t.opts.maxContent && t.opts.maxContent > 0 { ... }

File: widgets/text/text.go [hunk 3]
```diff
@@ -109,6 +109,17 @@ func (t *Text) Write(text string, wOpts ...WriteOption) error {
...
+			// If the new content is longer than all existing, simply reset
+			if len(string(r)) >= len(t.content) {
```
Comment (mum4k) [inline widgets/text/text.go:115]: I may be missing something, but since `r` is a single rune in this for loop, doesn't the expression `len(string(r))` always return 1?

File: widgets/text/text.go [hunk 4]
```diff
@@ -109,6 +109,17 @@ func (t *Text) Write(text string, wOpts ...WriteOption) error {
 		t.reset()
 	}
 	for _, r := range text {
+		// Implement basic buffer limit to avoid excess memory issues
+		if len(t.content) > 1 && len(t.content) >= t.opts.maxContent && t.opts.maxContent > 0 {
+			// If the new content is longer than all existing, simply reset
```
Comment (mum4k) [inline widgets/text/text.go:114]: How do we handle the case when a single `text` given to `Write` is longer than the configured `MaxContent`?

File: widgets/text/text_test.go [hunk 5]
```diff
@@ -809,6 +809,66 @@ func TestTextDraws(t *testing.T) {
...
+			},
+		},
+		{
+			desc:   "tests maxContent length not being limited",
+			canvas: image.Rect(0, 0, 72, 1),
+			opts: []Option{
+				RollContent(),
+			},
+			writes: func(widget *Text) error {
+				return widget.Write("1234567890abcdefghijklmnopqrstuvwxyz1234567890abcdefghijklmnopqrstuvwxyz")
```
Comment (mum4k) [inline widgets/text/text_test.go:839]: Thank you for including test coverage.

The selected test data (content) is fairly long, which makes it harder to reason about what they are testing. Can we minimize its length to the base cases? (a content of 1, max content set to 1, etc...). Using shorter content and MaxContent should arguably be the same behavior as with longer values, but the tests should be easier to follow and verify.

Once we do that, let's make sure we cover all the cases:
- adding content and fitting within MaxContent exactly
- adding content and overwriting portion of the buffer
- adding content and overwriting the full buffer
- adding content that is itself longer than MaxContent

This is the full list that comes to mind. On top of this we should keep the test with newlines and also add at least one test that sets MaxContent and uses unicode characters that are wider than a single cell.

Please let me know if I can provide additional clarification.

**Review by mum4k — CHANGES_REQUESTED** [review]
Thank you for the contribution @timbutler, leaving a few questions.

File: widgets/text/options.go (hunk 1 above)
Comment (timbutler) [reply widgets/text/options.go]: I've increased the commenting but it may require further clarification here. If my read of the code is correct, `t.content` is the cells so the limits should be based on the cells?

File: widgets/text/text.go (hunk 2 above)
Comment (timbutler) [reply widgets/text/text.go]: I've fixed this, you are correct in that it isn't required.

File: widgets/text/text.go (hunk 3 above)
Comment (timbutler) [reply widgets/text/text.go]: I've moved the logic to be outside of the stepping through the runes within the string so it should make more sense again.

File: widgets/text/text.go (hunk 4 above)
Comment (timbutler) [reply widgets/text/text.go]: I've now fixed this case.

File: widgets/text/text_test.go (hunk 5 above)
Comment (timbutler) [reply widgets/text/text_test.go:839]: I'd simply gone for a longer string for testing just so it stood out and had some variance. As you've pointed out through, it's probably not helpful anyway.

I'll increase the test coverage and commit once ready, I agree that there's extra scenarios to test.

Thanks again for the feedback!

Comment (mum4k) [reply widgets/text/text_test.go:839]: Happy to provide the feedback, the thanks goes to you for the contribution!

Comment (timbutler) [reply widgets/text/text_test.go]: Unicode chars did force a bit of a rethink of the logic to perfect, but the code commit a few days ago should now be correct and the tests (hopefully) enough to cover all cases.

The CI test failure doesn't appear to be related to my changes, please let me know if there's anything else required here.

File: widgets/text/options.go (hunk 1 above)
Comment (mum4k) [reply widgets/text/options.go]: That is correct, `t.content` is indeed in cells. How would you feel about renaming the option and its default to make it obvious? Something like:

`MaxContentCells` and `DefaultMaxContentCells`.

We could also mention that the number represents the maximum amount of terminal cells the content can take and note that some runes might take multiple cells for clarity.

File: widgets/text/text.go [hunk 6]
```diff
@@ -108,6 +109,25 @@ func (t *Text) Write(text string, wOpts ...WriteOption) error {
 	if opts.replace {
 		t.reset()
 	}
+
+	// If set, limit the buffer (if maxContent has been set)
+	if len(t.content) >= t.opts.maxContent && t.opts.maxContent > 0 {
+		fmt.Printf("len: %d strlen: %d", len(t.content), len(t.content))
```
Comment (mum4k) [inline widgets/text/text.go:115]: Was this a debug print that we want to remove before we merge this in?

File: widgets/text/text.go [hunk 7]
```diff
@@ -108,6 +109,25 @@ func (t *Text) Write(text string, wOpts ...WriteOption) error {
 	if opts.replace {
 		t.reset()
 	}
+
+	// If set, limit the buffer (if maxContent has been set)
```
Comment (mum4k) [inline widgets/text/text.go:113]: This latest version looks great, thank you for iterating on it. I wonder if we could make the Write function more readable by extracting all code related to content truncation into its own method. Something like:

```go
// Truncates the content and the new text so that the configured MaxContentCells
// is enforced after a Write. 
func (t *Text) truncateToMaxCells(text string) { ... }
```

The `Write` method would then just conditionally call the `truncateToMaxCells` method if `opts.maxContentCells` was set. This might make the "business logic" of `Write` easier to follow.

```go
if t.opts.maxContent > 0 {
  t.truncateToMaxCells(text)
}
```

File: widgets/text/text.go [hunk 8]
```diff
@@ -108,6 +109,25 @@ func (t *Text) Write(text string, wOpts ...WriteOption) error {
...
+	if len(t.content) >= t.opts.maxContent && t.opts.maxContent > 0 {
+		fmt.Printf("len: %d strlen: %d", len(t.content), len(t.content))
+		// If the new content is longer than all existing, simply reset
+		if len(text) >= len(t.content) {
+			t.reset()
+		} else {
+			// Truncate by the space required for the new entry
+			t.content = t.content[runewidth.StringWidth(text):]
```
Comment (mum4k) [inline widgets/text/text.go:121]: The calls to `runewidth.StringWidth` do take some work (at the very least O(n) scan over the text. Any way we could refactor this so that the code calls `runewidth.StringWidth(text)` only once?

File: widgets/text/text.go [hunk 9]
```diff
@@ -108,6 +109,25 @@ func (t *Text) Write(text string, wOpts ...WriteOption) error {
 	if opts.replace {
 		t.reset()
 	}
+
+	// If set, limit the buffer (if maxContent has been set)
+	if len(t.content) >= t.opts.maxContent && t.opts.maxContent > 0 {
```
Comment (mum4k) [inline widgets/text/text.go:114]: What do we do when someone sets the option to zero cells? Should that be an error when the option is being set?

Can we also add a test case?

File: widgets/text/text_test.go [hunk 10]
```diff
@@ -809,6 +809,112 @@ func TestTextDraws(t *testing.T) {
...
+					c,
+					"12345",
+					image.Point{0, 0},
+				)
+				testcanvas.MustApply(c, ft)
+				return ft
+			},
+		},
+		{
+			desc:   "tests maxContent partial bufffer replacement",
```
Comment (mum4k) [inline widgets/text/text_test.go:856]: Typo: "bufffer".

File: widgets/text/text_test.go [hunk 11]
```diff
@@ -809,6 +809,112 @@ func TestTextDraws(t *testing.T) {
...
+				c := testcanvas.MustNew(ft.Area())
+				// \n still counts as a chacter in the string length
+				testdraw.MustText(c, "ine3", image.Point{0, 0})
+				testdraw.MustText(c, "line4", image.Point{0, 1})
+				testcanvas.MustApply(c, ft)
+				return ft
+			},
+		},
+		{
+			desc:   "tests maxContent exact length of 5",
```
Comment (mum4k) [inline widgets/text/text_test.go:833]: Other interesting test cases are:

- length is just one under the set max
- length is just one over the set max
- multiple writes, the last write causes the content to be reset because the provided text is longer than the configured max
- multiple writes, the last writes truncates the existing content before writing
- multiple writes, the last write truncates text but leaves content untouched

We can also take a look at the coverage report to see which lines aren't covered.

Comment (mum4k) [inline widgets/text/text_test.go:833]: We should also consider what happens with unicode characters that are multi-cell and add a test case where the max content "cuts a rune in half" to verify that Termdash doesn't attempt to print half of a wide rune, but correctly skips it.

Unicode never fails to provide additional entertainment.

**Review by mum4k — CHANGES_REQUESTED** [review]
Looks great, left a few more comments and looks like we could still add more test cases.

Please feel free to let me know if you would prefer to merge this in and have me implement changes based on any of the listed comments. I am happy to help.

Fyi, the CI failure was indeed unrelated to this PR and should now be resolved (after you merge the devel branch).

File: widgets/text/text.go (hunk 6 above)
Comment (timbutler) [reply widgets/text/text.go:115]: Yes, just debug which crept in.

File: widgets/text/text.go (hunk 7 above)
Comment (timbutler) [reply widgets/text/text.go:113]: Happy for this to be moved out. Happy if you wanted to make these changes, the incoming `text` also needs to be truncated if required as well.

File: widgets/text/text.go (hunk 8 above)
Comment (timbutler) [reply widgets/text/text.go:121]: Unless my logic is bad, this should only be called twice during the whole Write procedure. I haven't profiled how intensive it is so if it's worth calling once then happy for this to be refactored.

File: widgets/text/text.go (hunk 9 above)
Comment (timbutler) [reply widgets/text/text.go:114]: Easiest fix is probably to set the conditional logic to >= 0. It'll then be ignored since the only valid size is 1 or greater.

File: widgets/text/text_test.go (hunk 11 above)
Comment (timbutler) [reply widgets/text/text_test.go:833]: I believe this is covered in the test `tests maxContent partial bufffer replacement`, as it has to truncate based on a character not the data (hence starting with 你 for the result). I was using this as a test case for this as I've never worked with Go for unicode before!

If I get the time I'll make some of the adjustments later in the week, otherwise happy if you want to merge and correct as well.

**Comment by mum4k** [comment]
Thank you very much @timbutler, I will be happy to merge as is and work on the remaining items.

Before we do that I have one more CLA related favor to ask for. Can you please explicitly state that you give me permission to merge this code into the master branch? This is required to pass the CLA check when I will be releasing the commits authored by you.

**Comment by timbutler** [comment]
Hi @mum4k, sorry for the delay. I'm happy for all of my code to be merged into the master branch.

**Comment by mum4k** [comment]
Thank you again for the confirmation and the contribution @timbutler.

**Review by mum4k — APPROVED** [review]
